cpp/core/memory/ReservationListener.h

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cstdlib>
+
+#include "jni/JniCommon.h"
+
+namespace gluten {
+
+class ReservationListener {
+ public:
+  static std::shared_ptr<ReservationListener> noop();
+
+  virtual ~ReservationListener() = default;
+
+  // Value of diff can be either positive or negative
+  virtual void reservationChanged(int64_t diff) = 0;
+
+ protected:
+  ReservationListener() = default;
+};
+
+class SparkReservationListener final : public ReservationListener {
+ public:
+  SparkReservationListener(
+      JavaVM* vm,
+      jobject jListenerLocalRef,
+      jmethodID jReserveMethod,
+      jmethodID jUnreserveMethod,
+      int64_t blockSize)
+      : vm_(vm), jReserveMethod_(jReserveMethod), jUnreserveMethod_(jUnreserveMethod), blockSize_(blockSize) {
+    JNIEnv* env;
+    attachCurrentThreadAsDaemonOrThrow(vm_, &env);
+    jListenerGlobalRef_ = env->NewGlobalRef(jListenerLocalRef);
+  }
```

Comment:
Why moving this out from `core/jni`?

`SparkReservationListener` relies on JNI it was in `core/jni`.

Reply:
ReservationListener concept was connected with memory management, so I think move to memory folder would make more sense, core/memory could depends on core/jni right? these are not independent.

Reply:
> core/memory could depends on core/jni right?

Can you elaborate? I thought they were independent. Thanks.

Reply:
I got it, will fix later.

Reply:
already revert.